torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
+        target_mbs: Optional[List] = None,
+        losses: Optional[List] = None,
+    ):
+        """
+        Operate on the microbatches using the interleaved 1f1b schedule (https://arxiv.org/pdf/2104.04473.pdf).
 
-        # Make sure all sends are finished
-        for work in sends_to_wait:
-            work.wait()
+        TODO: Interleaved 1F1B does not use sorted_batch_isend_irecv(). As a result, this schedule does
+        not support models with skip connections.
+        """
+        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)
 
+        # Based on the plan in Step 1 created in __init__:
+        # 2. Perform communication based on the pipeline_order
+        stage_index_to_stage: Dict[int, _PipelineStageBase] = {
+            stage.stage_index: stage for stage in self._stages
+        }
+        prev_rank: int = (self.rank - 1) % self.pp_group_size
+        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        for time_step, action in enumerate(self.pipeline_order[self.rank]):
+            prev_rank_ops = self.pipeline_order[prev_rank]
+            next_rank_ops = self.pipeline_order[next_rank]
+            ops: List[dist.P2POp] = []
+            if action is not None:
+                computation_type, mb_index, stage_index = action
+                if computation_type == _ComputationType.FORWARD:
+                    # perform forward computation
+                    stage = stage_index_to_stage[stage_index]
+                    output = stage.forward_one_chunk(
+                        arg_mbs[mb_index], kwarg_mbs[mb_index]
+                    )
+                    self._maybe_compute_loss(stage, output, target_mbs, mb_index)
+                    ops.extend(stage.get_fwd_send_ops())
+                elif computation_type == _ComputationType.BACKWARD:
+                    # perform backward computation
+                    stage = stage_index_to_stage[stage_index]
+                    loss = self._maybe_get_loss(stage, mb_index)
+                    stage.backward_one_chunk(loss=loss)
+                    if stage_index == 4 and mb_index == 4:
+                        pass
+                    ops.extend(stage.get_bwd_send_ops())
+                else:
+                    raise ValueError(f"Unknown computation type {computation_type}")
+
+            # look at neighboring ranks to see if I need to include any recv communication
+            if time_step < len(prev_rank_ops):
+                prev_rank_action = prev_rank_ops[time_step]
+            else:
+                prev_rank_action = None
+            if prev_rank_action is not None:
+                computation_type, mb_index, stage_index = prev_rank_action
+                # only handle sends for the forward from a previous rank
+                if computation_type == _ComputationType.FORWARD:
+                    if stage_index != self._num_stages - 1:
+                        stage = stage_index_to_stage[stage_index + 1]
```

Comment:
Note: this line is assuming that `stage_index + 1` is the receiver of `stage_index`. 
This is correct for the most cases, but also overrides the comm relationship returned by `stage.get_fwd_recv_ops()` (which will return the source rank based on pipe construction).

We should at least document the assumption here.